File: android/app/src/main/java/org/fs/news/model/intent/LoadSourceIntent.kt
/*
 *  Copyright (C) 2019 Fatih, News Stand Android Kotlin.
 *  
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.fs.news.model.intent

import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.fs.architecture.mvi.common.*
import org.fs.news.model.SourceModel
import org.fs.news.model.entity.Source
import org.fs.news.net.model.Resource
import org.fs.news.net.repo.SourceRepository
import org.fs.news.util.Operations.Companion.REFRESH
import java.io.IOException

class LoadSourceIntent(private val sourceRepository: SourceRepository): ObservableIntent<SourceModel>() {

  override fun invoke(): Observable<Reducer<SourceModel>> = sourceRepository.sources()
    .concatMap(::success)
    .onErrorResumeNext(::failure)
    .startWith(initial())
    .subscribeOn(Schedulers.io())

  private fun success(resource: Resource<List<Source>>): Observable<Reducer<SourceModel>> = when(resource) {
    is Resource.Success<List<Source>> -> Observable.just(
      { o -> o.copy(state = Operation(REFRESH), data = resource.data ?: emptyList()) },
      { o -> o.copy(state = Idle, data = emptyList()) })
    is Resource.Failure<List<Source>> -> Observable.just(
      { o -> o.copy(state = Failure(IOException(resource.message))) },
      { o -> o.copy(state = Idle) })
  }

  private fun failure(error: Throwable): Observable<Reducer<SourceModel>> = Observable.just(
    { o -> o.copy(state = Failure(error)) },
    { o -> o.copy(state = Idle) })

  private fun initial(): Reducer<SourceModel> = { o -> o.copy(state = Operation(REFRESH), data = emptyList()) }
}

File: ios/NewsStand/Model/Intent/LoadSourceIntent.swift
//
//  LoadSourceIntent.swift
//  NewsStand
//
//  Created by Fatih Şen on 29.06.2019.
//  Copyright © 2019 Fatih Şen. All rights reserved.
//

import Foundation
import RxSwift
import MVICocoa

class LoadSoruceIntent: ObservableIntent<SourceModel> {
	
	private let sourceRepository: SourceRepository
	
	init(sourceRepository: SourceRepository) {
		self.sourceRepository = sourceRepository
	}
	
	override func invoke() -> Observable<(SourceModel) -> SourceModel> {
		return sourceRepository.sources()
			.concatMap(success(_ :))
			.catchError(failure(_:))
			.startWith(initial())
			.delay(RxTimeInterval.milliseconds(500), scheduler: MainScheduler.asyncInstance)
			.subscribeOn(MainScheduler.asyncInstance)
	}
	
	private func initial() -> Reducer<SourceModel> {
		return { o in o.copy(state: .operationOf(refresh), data: []) }
	}
	
	private func success(_ resource: Resource<[Source]>) -> Observable<Reducer<SourceModel>> {
		switch resource {
			case .success(let data, _): return Observable.of(
				{ o in o.copy(state: .operationOf(refresh), data: data ?? [])},
				{ o in o.copy(state: .idle, data: []) })
			case .failure(_, let message): return Observable.of(
				{ o in o.copy(state: .failure(MVIError.of(description: message ?? ""))) },
				{ o in o.copy(state: .idle) })
		}
	}
	
	private func failure(_ error: Error) -> Observable<Reducer<SourceModel>> {
		return Observable.of(
			{ o in o.copy(state: .failure(error)) },
			{ o in o.copy(state: .idle) })
	}
}
